internal/literals/swap.go

```diff
@@ -0,0 +1,173 @@
+package literals
+
+import (
+	"go/ast"
+	"go/token"
+	"strconv"
+)
+
+type swap struct{}
+
+// check that the obfuscator interface is implemented
+var _ obfuscator = swap{}
+
+// Source: https://golang.org/ref/spec#Numeric_types
+const maxUInt8 = 255
+const maxUint16 = 65535
+
+func getIndexType(dataLen int) string {
+	switch {
+	case dataLen <= maxUInt8:
+		return "byte"
+	case dataLen <= maxUint16:
+		return "uint16"
+	default:
+		return "uint32"
+	}
+}
+
+func indexesToSlice(data []int) *ast.CompositeLit {
+	arr := &ast.CompositeLit{
+		Type: &ast.ArrayType{
+			Len: &ast.Ellipsis{}, // Performance optimization
+			Elt: ident(getIndexType(len(data))),
+		},
+		Elts: []ast.Expr{},
+	}
+	for _, data := range data {
+		arr.Elts = append(arr.Elts, intLiteral(strconv.Itoa(data)))
+	}
+	return arr
+}
+
+func (x swap) obfuscate(data []byte) *ast.BlockStmt {
+	maxJunkIdxCount := len(data) / 2
+	if maxJunkIdxCount == 0 {
+		maxJunkIdxCount = 1
+	}
+	count := len(data) + genRandIntn(maxJunkIdxCount)
+	if count%2 != 0 {
+		count++
+	}
+	shiftKey := byte(genRandIntn(maxUInt8))
+
+	indexes := generateIntSlice(len(data), count)
+	for i := len(indexes) - 2; i >= 0; i -= 2 {
```

Comment:
I think you should start at zero (or two) and count up to len(indexes)

Reply:
No, the reverse order is important for correct processing of the swap chain (1<->9 - 1<->3 - 3<->6).

Reply:
Ok